text/0000-legal-double-reference.md

```diff
@@ -0,0 +1,88 @@
+
+- Feature Name: Legal Double Reference
+- Start Date: 2017-12-27
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The Rust borrow checker does not allow you to make an immutable reference to something while a mutable reference exists. The reasons for this are made clear, however its implementation is too strict; I propose that an immutable reference to something should be allowed to be created while there is an existing mutable reference if it is dropped before the next use of the mutable reference.  
+This change will mean that getting useful values from types while mutating them will be less difficult and encourage functional programming in Rust while preserving its safety.
+
+# Motivation
+[motivation]: #motivation
+
+This change should be made because it encourages use of existing functional methods in Rust which can sometimes be difficult to use.  
+The use cases of this change are times when a value is being mutated but during that time some method needs to be called which requires an immutable reference to the value.
+```rust
+let mut a = vec![0u32; 3];
+
+a.iter_mut()
+.for_each(|b| {
+    //Do stuff with b.
+    let c = a.len(); //E0502
+    //Do more stuff with b.
+});
+```

```

Comment:
This snippet has far greater issues than `a` being mutably borrowed *inside the closure*:

1. Closures begin borrowing from their environment as soon as they are created, not just when they are called.  In other words, calling `a.len()` inside the closure does not cause it to be immutably borrowed each iteration.  It is immutably borrowed *once,* and this borrow lasts for the entire duration of the `for_each` loop.
2. A function like `iter_mut` is currently allowed to do things that put `a` in an invalid state (i.e. where `len()` could invoke undefined behavior), because it knows that it has unique access to that object until the `IterMut` struct is dropped.